pkg/discovery/kubernetes/control-point-discovery.go

```diff
@@ -0,0 +1,178 @@
+					if slices.Contains(scalableResources, apiResource.Name) {
+						groupVersion, parseErr := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+						if parseErr != nil {
+							log.Error().Err(parseErr).Msg("Unable to parse group version")
+							return parseErr
+						}
+						groupVersionResource := schema.GroupVersionResource{
+							Group:    groupVersion.Group,
+							Version:  groupVersion.Version,
+							Resource: apiResource.Name,
+						}
+						// Add to groupVersionResourceSet
+						groupVersionResourceSet[groupVersionResource] = apiResource.Kind
+					}
+				}
+			}
+
+			// Track each scalable resource
+			for groupVersionResource, kind := range groupVersionResourceSet {
+				log.Info().Msgf("Starting watch for Group: %s, Version: %s, Resource: %s", groupVersionResource.Group, groupVersionResource.Version, groupVersionResource.Resource)
+				resourceInterface := cpd.dynamicClient.Resource(groupVersionResource)
+
+				// watch for changes
+				_, controller := cache.NewInformer(
+					&cache.ListWatch{
+						ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+							return resourceInterface.List(cpd.ctx, options)
+						},
+						WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+							return resourceInterface.Watch(cpd.ctx, options)
+						},
+					},
+					&unstructured.Unstructured{},
+					0,
+					cpd.createResourceEventHandlerFuncs(groupVersionResource, kind),
+				)
+
+				// start controller
+				panichandler.Go(func() {
+					controller.Run(cpd.ctx.Done())
+				})
+			}
+
+			<-cpd.ctx.Done()
+			return nil
+		}
+
+		boff := backoff.NewConstantBackOff(5 * time.Second)
+		_ = backoff.Retry(operation, backoff.WithContext(boff, cpd.ctx))
+
+		log.Info().Msg("Stopped kubernetes control point watcher")
+	})
+}
+
+func (cpd *controlPointDiscovery) createResourceEventHandlerFuncs(groupVersionResource schema.GroupVersionResource, kind string) cache.ResourceEventHandlerFuncs {
+	controlPointFromObject := func(obj interface{}) ControlPoint {
+		// read the name of the resource
+		name := obj.(*unstructured.Unstructured).GetName()
+		namespace := obj.(*unstructured.Unstructured).GetNamespace()
+		return ControlPoint{
+			Group:     groupVersionResource.Group,
+			Version:   groupVersionResource.Version,
+			Kind:      kind,
+			Name:      name,
+			Namespace: namespace,
+		}
+	}
+
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			controlPoint := controlPointFromObject(obj)
+			cpd.controlPointStore.Add(controlPoint)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			controlPoint := controlPointFromObject(obj)
+			cpd.controlPointStore.Update(controlPoint)
+		},
+		DeleteFunc: func(obj interface{}) {
+			controlPoint := controlPointFromObject(obj)
+			cpd.controlPointStore.Delete(controlPoint)
+		},
+	}
+}
+
+func (cpd *controlPointDiscovery) stop() {
+	cpd.cancel()
+	cpd.waitGroup.Wait()
```

Comment:
Question : 
The intention of stop is to stop all the goroutines or wait until the goroutine work is done ?

Reply:
The intention is for Fx to detect if a goroutine is stuck at shutdown over [here](https://github.com/fluxninja/aperture/blob/tgill/k8s-control-point-discovery/pkg/discovery/kubernetes/provide.go#L93)